### PR TITLE
Massive Update: Convert BMS IC code into standard Zephyr driver

### PR DESCRIPTION
**Pull request by martinjaeger**
This fairly large PR converts the existing BMS IC drivers into a proper Zephyr driver. It was not possible to split the commit into several steps, as all drivers and tests had to be converted at once in order to pass CI.

## Rationale

The previous implementation was still based on the Mbed-based firmware and did not follow the Zephyr driver conventions.

Advantages of using the Zephyr driver API:

- Better abstraction and support for multiple ICs on one BMS board.
- Cleaner and more advanced configuration through Devicetree.
- Cleaner implementation of unit tests, as we can use the I2C emulator drivers for mocking.
- This repo can be used as a module for custom BMS implementations.

## Implementation Details

The previous implementation did not clearly separate the functions related to the BMS IC (bq769x2, etc.) and the higher-level BMS functions like SOC calculation, state machine or cell parameter setup. All functions were prefixed with `bms_`.

The new implementation introduces a generic `bms_ic` driver for chips such as the bq769x0, bq769x2 and isl94202. Analog Devices ICs (LTC68xx series and ADBMSxxxx) can be added in the future and were already considered in the API design, which can be found in `include/drivers/bms_ic.h`.

Note: These chips are sometimes also called analog front ends or battery monitors. However, especially the bq76952 can do much more than just monitoring or reading analog values. I think a generic name `bms_ic` is better than `bms_afe` or `bat_monitor`.

BMS higher-level functions are defined in `include/bms/bms.h` and prefixed just with `bms_`. The file `include/bms/bms_common.h` contains functions and definitions used both by the BMS IC driver and the BMS application code.

## Testing

The code has been extensively tested with the bq76952 (BMS C1 board). Testing for other chips is still ongoing.

### CODE REVIEW

**Review by pasrom — CHANGES_REQUESTED** [review]
I flashed this changes to the BMS C1 v0.3 and looked at the UART & CAN Communication which worked as expected.

**Error Handling:** General, if possible, I would use specific error codes (like -EPERM) instead of return -1 for clearer error indication. This aligns better with our coding standards and aids in debugging.
**Commit Structure:** For future commits, I find it really helpful if renamings, like changing `&bms.conf.dis_sc_limit` to `&bms.ic_conf.dis_sc_limit`, are done in separate commits. This approach helps keep our commit history clean and easy to follow. It's a good practice that enhances clarity for everyone.

File: app/src/data_objects.cpp [hunk 1]
```diff
@@ -8,37 +8,34 @@
...
 
-#include "board.h"
-
-#include "bms.h"
+#include <bms/bms.h>
 
 #include <stdio.h>
 
-extern Bms bms;
+static const struct device *bms_ic = DEVICE_DT_GET(DT_ALIAS(bms_ic));
```
[the remark this answers is not in the document]
Comment (martinjaeger) [reply app/src/data_objects.cpp]: Good point. Fixed.

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2021 Nordic Semiconductor ASA
```
[the remark this answers is not in the document]
Comment (martinjaeger) [reply CMakeLists.txt:1]: For the CMakeLists.txt it's probably not really relevant (as it's a very simple file). However, the other files with Nordic copyright notice were copied from the Zephyr example repo. The Apache license requires you to keep the original copyright, so I did that.

One could argue that Nordic would also become a "Libre Solar Project Contributor", but as they kind of contributed passively, I think it's more correct to keep the original copyright notice in this case.

File: app/src/data_objects.cpp [hunk 3]
```diff
@@ -177,85 +184,87 @@ THINGSET_ADD_ITEM_FLOAT(APP_ID_CONF_PRESET_LFP, APP_ID_CONF_PRESET_LFP_CAPACITY,
 
 THINGSET_ADD_GROUP(TS_ID_ROOT, APP_ID_MEAS, "Meas", THINGSET_NO_CALLBACK);
 
-THINGSET_ADD_ITEM_FLOAT(APP_ID_MEAS, APP_ID_MEAS_PACK_VOLTAGE, "rPackVoltage_V",
```
[the remark this answers is not in the document]
Comment (martinjaeger) [reply app/src/data_objects.cpp:180]: Yes and no. As the ThingSet protocol is self-explanatory, e.g. the smartphone app would just get the new name and display that. However, a database entry in the backend may change.

I reverted this change, as it's not relevant to this PR.

File: app/src/data_objects.cpp [hunk 4]
```diff
@@ -177,85 +184,87 @@ THINGSET_ADD_ITEM_FLOAT(APP_ID_CONF_PRESET_LFP, APP_ID_CONF_PRESET_LFP_CAPACITY,
...
+THINGSET_ADD_ITEM_FLOAT(APP_ID_MEAS, APP_ID_MEAS_SOC, "rSOC_pct", &bms.soc, 1, THINGSET_ANY_R,
+                        TS_SUBSET_LIVE);
 
 THINGSET_ADD_ITEM_UINT32(APP_ID_MEAS, APP_ID_MEAS_ERROR_FLAGS, "rErrorFlags",
-                         &bms.status.error_flags, THINGSET_ANY_R, TS_SUBSET_LIVE);
+                         &bms.ic_data.error_flags, THINGSET_ANY_R, TS_SUBSET_LIVE);
 
-THINGSET_ADD_ITEM_UINT16(APP_ID_MEAS, APP_ID_MEAS_BMS_STATE, "rBmsState", &bms.status.state,
-                         THINGSET_ANY_R, TS_SUBSET_LIVE);
+THINGSET_ADD_ITEM_UINT8(APP_ID_MEAS, APP_ID_MEAS_BMS_STATE, "rBmsState", (uint8_t *)&bms.state,
```
[the remark this answers is not in the document]
Comment (martinjaeger) [reply app/src/data_objects.cpp:217]: Yes, it's relevant, as the state was changed from uint16 to an enum as part of the driver rework.

File: drivers/bms_ic/Kconfig
[the remark this answers is not in the document]
Comment (martinjaeger) [reply drivers/bms_ic/Kconfig:1]: Good point. Done.

File: dts/bindings/bms_ic/ti,bq769x2-i2c.yaml [hunk 5]
```diff
@@ -2,8 +2,8 @@
 # SPDX-License-Identifier: Apache-2.0
 
 description: |
-    Texas Instruments BQ769x2 3s to 16s High Accuracy Battery Monitor and
-    Protector for Li-Ion, Li-Polymer, and LiFePO4 Battery Packs (SPI)
+  Texas Instruments BQ769x2 3s to 16s High Accuracy Battery Monitor and
+  Protector for Li-Ion, Li-Polymer, and LiFePO4 Battery Packs (I2C)
```
[the remark this answers is not in the document]
Comment (martinjaeger) [reply dts/bindings/bms_ic/ti,bq769x2-i2c.yaml]: Yep, was kind of a drive-by clean-up, which should have gone into a separate commit. I have reverted it for now.

**Comment by martinjaeger** [comment]
Thanks @pasrom for your review.

As a note regarding atomic commits: You are absolutely right that changes should be done in small atomic steps, so that they are easier to understand and review I'm normally trying to follow this practice as well. However, in this case the driver for the bq769x2 was developed in another project and I could not cherry-pick the individual steps from that project, as the file structure was slightly different and the steps did not cover the drivers for bq769x0 and isl94202. In order to avoid a huge amount of extra work to artificially replicate the steps again, I chose to squash everything into one commit.

**Review by pasrom — APPROVED** [review]
@martinjaeger, thanks for the explanations and clarifications. The changes look good to me now. Approved!

**Pull request merged**